### PR TITLE
add ECN events

### LINKED ISSUE #212: add ECN events
Endpoints using ECN will perform validation of ECN, as described in https://datatracker.ietf.org/doc/html/rfc9000#section-13.4.2. There are a few different conditions that can lead to a failure of ECN validation, which in turns leads to disabling of ECN for that path. We should have an event that tells us _that_ ECN validation failed and _why_.

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #212.

### CODE REVIEW

**Comment by marten-seemann** [comment]
2023-01-05 conversation:
* should this be on the datagram_ instead of packet_ event?
* is there a default value? Should this be a separate event like the spin bit?
* use semantic names?

**Comment by marten-seemann** [comment]
I've updated the PR to:
* use the semantic names instead of the two bits
* defined Non-ECT as the default value
* moved this to the datagram_ events

While technically more correct, I'm not a big fan of the last one, as the datagram_ events have an importance of "Extra", so it's more likely ECN won't be logged. For example, quic-go currently doesn't use the datagram_ events at all.

**Comment by rmarx** [comment]
I think that's a fair point about `datagrams_` vs `packet_`. IIUC, the only place where this matters anyway is during the handshake, as there can be only a single 1-RTT packet in one datagram. I'm not sure if/how much ECN comes into play during the handshake?

I think the argument could be made that for the handshake, the ECN bits could be reflected on/copied to each of the potentially coalesced QUIC `packet_` events. They can still be tracked to the singular datagram they were in using the `datagram_id` construction after all. 

It's a bit less "clean"/"correct", but that can be said of many things in qlog ;) What do you think, @LPardue?

**Comment by marten-seemann** [comment]
We could also put it on both event types, and allow implementations that don’t log datagram events to put it in the packet. Of course, this shifts some complexity to consumers of the qlog.

**Comment by marten-seemann** [comment]
I added events for the ECN testing state machine from RFC 9000, A.4. This should now cover the (non-L4S) use case of ECN.

Would appreciate a review of this PR.

**Comment by rmarx** [comment]
This looks good to me overall, but I wonder if it would be better to make the ECN enum extensible as well (to allow for easy expansion into L4S in the future if needed):

e.g., something like

```
$ECNfield /= {
    * text => any
}

ECN = "Non-ECT" / "ECT(1)" / "ECT(0)" / "CE"

$ECNfield /= ECN
```

Additionally, I wonder if we want some way to remove some chattiness here... I imagine that most values in DatagramsSent will be identical once the probing is done, and that you really only want to log when a CE is received in DatagramsReceived. Not sure if it's needed, but maybe some text to state that explicitly like "no need to log with EVERY packet, only if values change from previously logged value"? (though arguably you could say that's also observable through the reflected values in the ACKs...).

**Comment by marten-seemann** [comment]
> This looks good to me overall, but I wonder if it would be better to make the ECN enum extensible as well (to allow for easy expansion into L4S in the future if needed):
> 
> e.g., something like
> 
> ```
> $ECNfield /= {
>     * text => any
> }
> 
> ECN = "Non-ECT" / "ECT(1)" / "ECT(0)" / "CE"
> 
> $ECNfield /= ECN
> ```

I don't think that's needed. There are only 4 ECN values (2 bits) allowed by the IP header. L4S doesn't change that.



> Additionally, I wonder if we want some way to remove some chattiness here... I imagine that most values in DatagramsSent will be identical once the probing is done, and that you really only want to log when a CE is received in DatagramsReceived. Not sure if it's needed, but maybe some text to state that explicitly like "no need to log with EVERY packet, only if values change from previously logged value"? (though arguably you could say that's also observable through the reflected values in the ACKs...).

It's not observable from the ACKs, the ACK only tells you what happened along the path. Not logging it on every packet sounds good though.

File: draft-ietf-quic-qlog-quic-events.md [hunk 1]
```diff
@@ -1561,6 +1599,14 @@ KeyType = "server_initial_secret" /
 ~~~
 {: #keytype-def title="KeyType definition"}
 
+## ECN
+~~~ cddl
+ECN = "Non-ECT" / "ECT(1)" / "ECT(0)" / "CE"
```
Comment (marten-seemann) [inline draft-ietf-quic-qlog-quic-events.md:1604]: See https://en.wikipedia.org/wiki/Explicit_Congestion_Notification#Operation_of_ECN_with_IP
```suggestion
ECN = "Not-ECT" / "ECT(1)" / "ECT(0)" / "CE"
```

File: draft-ietf-quic-qlog-quic-events.md [hunk 2]
```diff
@@ -1413,6 +1426,31 @@ RecoveryMarkedForRetransmit = {
...
+ECNStateUpdated = {
+   ? old: ECNState
+    new: ECNState
+}
+
+ECNState =
+  ; ECN testing not yet started, ECN support not determined yet.
+  "unknown" /
+  ; ECN testing in progress
+  "testing" /
```
Comment (marten-seemann) [inline draft-ietf-quic-qlog-quic-events.md:1445]: This is backwards. Testing comes first, and after that comes unknown.
```suggestion
  ; ECN testing in progress
  "testing" /
  ; ECN state unknown, waiting for acknowledgements for testing packets
  "unknown" /
```

**Comment by rmarx** [comment]
> It's not observable from the ACKs, the ACK only tells you what happened along the path.

not sure what you mean by this. The ACKs reflect the incoming ECN info to the sender, so looking at incoming ACKs shows you what happened from YOU -> THEM and looking at outgoing ACKs (the one's we're sending) should show what happened THEM -> US. Put differently: your outgoing ACKs should reflect what you saw in DatagramsReceived, no? (though, to be fair, maybe I'm misunderstanding how all of this works).

**Comment by marten-seemann** [comment]
> > It's not observable from the ACKs, the ACK only tells you what happened along the path.
> 
> not sure what you mean by this. The ACKs reflect the incoming ECN info to the sender, so looking at incoming ACKs shows you what happened from YOU -> THEM and looking at outgoing ACKs (the one's we're sending) should show what happened THEM -> US. Put differently: your outgoing ACKs should reflect what you saw in DatagramsReceived, no? (though, to be fair, maybe I'm misunderstanding how all of this works).

You're right, for the receive side. For the send side this doesn't work though.

**Comment by rmarx** [comment]
Ok, agreed then :) 
Though, as I said, for the send side the codepoints will (probably?) usually always be the same after detection/negotiation/testing (e.g., if capable, you always send ECT(0) from then on), so no need to keep logging after that. Hence the original comment above.

**Comment by marten-seemann** [comment]
That makes sense. I updated the definition such that the default is now the value on the last event of that type (c7cd9bbcf21e7065ef51c26cb726b5c870522221).

@LPardue @rmarx Would you mind taking another look at this PR? Would love to get it in before the draft cutoff date for Prague.

File: draft-ietf-quic-qlog-quic-events.md [hunk 3]
```diff
@@ -875,6 +878,12 @@ QUICDatagramsSent = {
     ; The RawInfo fields do not include the UDP headers,
     ; only the UDP payload
     ? raw: [+ RawInfo]
+
+    ; ECN bits in the IP header
+    ; if not set, defaults the value used on the last
```
Comment (LPardue) [inline draft-ietf-quic-qlog-quic-events.md:883]:
```suggestion
    ; if not set, defaults to the value used on the last
```

File: draft-ietf-quic-qlog-quic-events.md [hunk 4]
```diff
@@ -1413,6 +1428,31 @@ RecoveryMarkedForRetransmit = {
 ~~~
 {: #recovery-markedforretransmit-def title="RecoveryMarkedForRetransmit definition"}
 
+## ecn_state_updated {#recovery-ecnstateupdated}
+Importance: Extra
+
+This event indicates a progression in the ECN state machine as described in section
+A.4 of RFC 9000.
```
Comment (LPardue) [inline draft-ietf-quic-qlog-quic-events.md:1435]:
```suggestion
A.4 of {{QUIC-TRANSPORT}}.
```

**Review by LPardue — APPROVED** [review]
LGTM modulo 2 nits

**Pull request merged**